registry/auth/token/accesscontroller.go

```diff
@@ -102,13 +105,36 @@ func (ac authChallenge) Status() int {
 	return http.StatusUnauthorized
 }
 
+func buildAutoRedirectURL(r *http.Request, autoRedirectPath string, forceTLS bool) string {
+	scheme := "https"
+
+	if !forceTLS {
+		if r.TLS != nil {
+			scheme = "https"
+		} else if len(r.URL.Scheme) > 0 {
+			scheme = r.URL.Scheme
+		}
+
+		if forwardedProto := r.Header.Get("X-Forwarded-Proto"); len(forwardedProto) > 0 {
+			scheme = forwardedProto
+		}
```

Comment:
I'm finding it a bit hard to follow the logic here. Would you mind elaborating on what is it you are trying to achieve here?

To me, this reads as follows: if we are not explicitly enforcing TLS but the actual request is via TLS encrypted connection we actually do enforce it _unless_ we actually override it if `X-Fwd-Proto` header is set

Reply:
Yes, you are right. 

However, it is also plausible that the admin has not set the `X-Forwarded-Proto` header，should we return `http`? or is it imperative to have the `X-Forwarded-Proto header`?